hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/upgrade/HDDSLayoutVersionManager.java

```diff
@@ -31,12 +43,61 @@
 public class HDDSLayoutVersionManager extends
     AbstractLayoutVersionManager<HDDSLayoutFeature> {
 
+  private static final Logger LOG =
+      LoggerFactory.getLogger(HDDSLayoutVersionManager.class);
+  private static final String[] HDDS_CLASS_UPGRADE_PACKAGES = new String[]{
+      "org.apache.hadoop.hdds.server.scm",
+      "org.apache.hadoop.ozone.container"};
+
   public HDDSLayoutVersionManager(int layoutVersion) throws IOException {
     init(layoutVersion, HDDSLayoutFeature.values());
+    registerUpgradeActions(HDDS_CLASS_UPGRADE_PACKAGES);
   }
 
   public static int maxLayoutVersion() {
     HDDSLayoutFeature[] features = HDDSLayoutFeature.values();
     return features[features.length - 1].layoutVersion();
   }
+
+  /**
+   * Scan classpath and register all actions to layout features.
+   */
+  @VisibleForTesting
+  void registerUpgradeActions(String[] packageNames) {
```

Comment:
As a general practice classpath scan and reflective class loading is done from static methods so that classpath configuration errors can be thrown before initializing the application.

Reply:
Yes, I agree. But, in this case the classpath loading is more investigative. It does not need to find any upgrade action at all. If it does find one, we will register it.